### PR TITLE
Add support for kilosort>=4.0.12

### LINKED ISSUE #3050: KS4: initialize_ops() missing 1 required positional argument: 'save_preprocessed_copy'
Hi, everyone.

I had to update Kilosort4 library in my environment to solve a bug previously  reported in [Kilosort](https://github.com/MouseLand/Kilosort/issues?q=is%3Aissue). Now, it looks like they added a new argument in the [run_kilosort.py](https://github.com/MouseLand/Kilosort/blob/main/kilosort/run_kilosort.py) file. 

image

Now I am getting this error while using this code:
```
sorting = si.run_sorter(sorter_name = 'kilosort4', recording = recording, 
                        folder = Path('C:/Path/to/Folder/KS4_output'), 
                        do_correction = False, verbose = True, docker_image = False)
```

SpikeSortingError                         Traceback (most recent call last)
Cell In[8], line 3
      1 job_kwargs = dict(n_jobs=-1, chunk_duration='1s', progress_bar=True)
----> 3 sorting = si.run_sorter(sorter_name = 'kilosort4', recording = AP_section, 
      4                         folder = Path('C:/Users/Juan/Downloads/KS4_output'), 
      5                         do_correction = False, verbose = True, docker_image = False)

File ~\spikeinterface\src\spikeinterface\sorters\runsorter.py:185, in run_sorter(sorter_name, recording, folder, remove_existing_folder, delete_output_folder, verbose, raise_error, docker_image, singularity_image, delete_container_files, with_output, output_folder, **sorter_params)
    178             container_image = singularity_image
    179     return run_sorter_container(
    180         container_image=container_image,
    181         mode=mode,
    182         **common_kwargs,
    183     )
--> 185 return run_sorter_local(**common_kwargs)

File ~\spikeinterface\src\spikeinterface\sorters\runsorter.py:245, in run_sorter_local(sorter_name, recording, folder, remove_existing_folder, delete_output_folder, verbose, raise_error, with_output, output_folder, **sorter_params)
    243 SorterClass.set_params_to_folder(recording, folder, sorter_params, verbose)
    244 SorterClass.setup_recording(recording, folder, verbose=verbose)
--> 245 SorterClass.run_from_folder(folder, raise_error, verbose)
    246 if with_output:
    247     sorting = SorterClass.get_result_from_folder(folder, register_recording=True, sorting_info=True)

File ~\spikeinterface\src\spikeinterface\sorters\basesorter.py:292, in BaseSorter.run_from_folder(cls, output_folder, raise_error, verbose)
    289         print(f"{sorter_name} run time {run_time:0.2f}s")
    291 if has_error and raise_error:
--> 292     raise SpikeSortingError(
    293         f"Spike sorting error trace:\n{log['error_trace']}\n"
    294         f"Spike sorting failed. You can inspect the runtime trace in {output_folder}/spikeinterface_log.json."
    295     )
    297 return run_time

SpikeSortingError: Spike sorting error trace:
Traceback (most recent call last):
  File "C:\Users\Juan\spikeinterface\src\spikeinterface\sorters\basesorter.py", line 257, in run_from_folder
    SorterClass._run_from_folder(sorter_output_folder, sorter_params, verbose)
  File "C:\Users\Juan\spikeinterface\src\spikeinterface\sorters\external\kilosort4.py", line 197, in _run_from_folder
    ops = initialize_ops(settings, probe, recording.get_dtype(), do_CAR, invert_sign, device)
TypeError: initialize_ops() missing 1 required positional argument: 'save_preprocessed_copy'

I suppose it is necessary to add this new argument to your [kilosrt4.py](https://github.com/SpikeInterface/spikeinterface/blob/main/src/spikeinterface/sorters/external/kilosort4.py) file

Thank you for your help!

### PR DESCRIPTION
**Pull request by alejoe91**
Plus some safe casting in the InjectTemplateRecording

Fixes #3050 #3020 #3007

### CODE REVIEW

**Comment by samuelgarcia** [comment]
cool.

**Review by samuelgarcia — APPROVED** [review]

**Review by chrishalcrow — APPROVED** [review]
Looks reasonable and it's working on our data. 

Just curious, the verbose output looks worse than the previous version. Is there some reason for that? Doesn't look like much has changed on our end?

now:
Screenshot 2024-06-20 at 15 53 04

before:
Screenshot 2024-06-20 at 15 58 01

**Comment by JoeZiminski** [comment]
Hey @alejoe91, just to check my understanding is this also related to the `uint` bug from #2839 (+ [comment](https://github.com/SpikeInterface/spikeinterface/issues/2839#issuecomment-2110258359))? This is casting to `int` deep in the `get_traces` machinery?

**Comment by alejoe91** [comment]
> Hey @alejoe91, just to check my understanding is this also related to the `uint` bug from #2839? This is casting to `int` deep in the `get_traces` machinery?

It's currently doing it only for the generated recording, but we could cast in the main get_traces. I think there's no downside

**Comment by alejoe91** [comment]
> Looks reasonable and it's working on our data.
> 
> Just curious, the verbose output looks worse than the previous version. Is there some reason for that? Doesn't look like much has changed on our end?
> 
> now: Screenshot 2024-06-20 at 15 53 04
> 
> before: Screenshot 2024-06-20 at 15 58 01

@chrishalcrow that's on the Kilosort side. Without the logger now only progress bars are printed...

**Comment by alejoe91** [comment]
> > Hey @alejoe91, just to check my understanding is this also related to the `uint` bug from #2839? This is casting to `int` deep in the `get_traces` machinery?
> 
> It's currently doing it only for the generated recording, but we could cast in the main get_traces. I think there's no downside

I'll make a separate PR for this

**Review by JoeZiminski — COMMENTED** [review]
Look good! An extra note to the comments, `save_sorting` has a new parameter `save_preprocessed_copy` in v4.0.12. I don't know if we need to / want to explicitly support this at this stage.

I had a look through how the signature of all imported KS4 functions change across versions and they have been generally stable (except for changes addressed already in the SpikeInterface code). However, further to my below comment I think I would advocate using explicitly keyword arguments for all function calls. Although unlikely (but definitely not impossible over the timescale of years), some nasty bugs could be introduced if on the KS4 end arguments are: a) changed in order b) one removed and one added at the same time. We are lucky if they cause a crash but they could feasibly cause non-breaking changes to results 😱 .

File: src/spikeinterface/sorters/external/kilosort4.py [hunk 1]
```diff
@@ -194,11 +206,17 @@ def _run_from_folder(cls, sorter_output_folder, params, verbose):
         data_dir = ""
         results_dir = sorter_output_folder
         filename, data_dir, results_dir, probe = set_files(settings, filename, probe, probe_name, data_dir, results_dir)
-        ops = initialize_ops(settings, probe, recording.get_dtype(), do_CAR, invert_sign, device)
+        if version.parse(cls.get_sorter_version()) >= version.parse("4.0.12"):
+            ops = initialize_ops(settings, probe, recording.get_dtype(), do_CAR, invert_sign, device, False)
```
[the remark this answers is not in the document]
Comment (alejoe91) [reply src/spikeinterface/sorters/external/kilosort4.py:210]: Can you make a separate PR for this?

[the remark this answers is not in the document]
Comment (JoeZiminski) [reply src/spikeinterface/sorters/external/kilosort4.py:210]: Sure will do!

[the remark this answers is not in the document]
Comment (JoeZiminski) [reply src/spikeinterface/sorters/external/kilosort4.py:210]: Sure will do! /remind me in 3 days

**Pull request merged**